File: examples/latest/navigate/flutter/hiking_diary_app/lib/HikingApp.dart
/*
 * Copyright (C) 2023 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import 'dart:core';

import 'package:flutter/material.dart';
import 'package:here_sdk/animation.dart';
import 'package:here_sdk/core.dart';
import 'package:here_sdk/core.errors.dart';
import 'package:here_sdk/location.dart';
import 'package:here_sdk/mapview.dart';
import 'package:here_sdk/navigation.dart';
import 'package:hiking_diary_app/OutdoorRasterLayer.dart';
import 'package:hiking_diary_app/locationfilter/LocationFilterAbstract.dart';
import 'package:hiking_diary_app/positioning/HEREPositioningProvider.dart';
import 'package:hiking_diary_app/positioning/HEREPositioningVisualizer.dart';

import 'GPXManager.dart';
import 'MessageNotifier.dart';
import 'locationfilter/DistanceAccuracyLocationFilter.dart';

class HikingApp implements LocationListener, LocationStatusListener {
  late HEREPositioningProvider herePositioningProvider;
  late HerePositioningVisualizer positioningVisualizer;
  late GPXManager gpxManager;

  final BuildContext context;
  final HereMapController mapView;
  final MessageNotifier messageNotifier;

  MapPolyline? myPathMapPolyline;
  bool isHiking = false;
  bool isGPXTrackLoaded = false;
  bool moveMapToCurrentLocation = true;
  GPXTrackWriter gpxTrackWriter = GPXTrackWriter();
  OutdoorRasterLayer? outdoorRasterLayer;
  LocationFilterAbstract? locationFilter;

  HikingApp(this.context, this.mapView, this.messageNotifier) {
    herePositioningProvider = HEREPositioningProvider();
    locationFilter = DistanceAccuracyLocationFilter();
    gpxManager = GPXManager("myGPXDocument.gpx", context);
    positioningVisualizer = HerePositioningVisualizer(mapView);
    outdoorRasterLayer = OutdoorRasterLayer(mapView);

    Location? location = herePositioningProvider.getLastKnownLocation();
    if (location != null) {
      mapView.camera.lookAtPoint(location.coordinates);
    }
    herePositioningProvider.startLocating(this, LocationAccuracy.navigation);

    setMessage("** Hiking Diary **");
  }

  @override
  void onFeaturesNotAvailable(List<LocationFeature> features) {}

  @override
  void onStatusChanged(LocationEngineStatus locationEngineStatus) {}

  @override
  void onLocationUpdated(Location location) {
    if (moveMapToCurrentLocation) {
      moveMapToCurrentLocation = false;
      _animateCameraToCurrentLocation(location);
    }

    positioningVisualizer.updateLocationIndicator(location);
    if (isHiking) {
      positioningVisualizer.renderUnfilteredLocationSignals(location);
    }

    if (isHiking && locationFilter!.checkIfLocationCanBeUsed(location)) {
      gpxTrackWriter.onLocationUpdated(location);
      MapPolyline mapPolyline = _updateTravelledPath();
      var distanceTravelled = getLengthOfGeoPolylineInMeters(mapPolyline.geometry);
      setMessage("Hike Distance: " + distanceTravelled.toString() + " m");
    }
  }

  void clearMap() {
    if (myPathMapPolyline != null) {
      mapView.mapScene.removeMapPolyline(myPathMapPolyline!);
      myPathMapPolyline = null;
    }
    positioningVisualizer.clearMap();
  }

  // Load the selected diary entry and show the polyline related to that hike.
  void loadDiaryEntry(int index) {
    if (isHiking) {
      setMessage("Stop hiking first.");
      return;
    }

    isGPXTrackLoaded = true;

    // Load the hiking trip.
    GPXTrack? gpxTrack = gpxManager.getGPXTrack(index);
    if (gpxTrack == null) {
      return;
    }

    List<GeoCoordinates> diaryGeoCoordinatesList = gpxManager.getGeoCoordinatesList(gpxTrack);
    GeoPolyline diaryGeoPolyline;
    int distanceTravelled = 0;

    try {
      diaryGeoPolyline = new GeoPolyline(diaryGeoCoordinatesList);
      distanceTravelled = getLengthOfGeoPolylineInMeters(diaryGeoPolyline);

      _addMapPolyline(diaryGeoPolyline);
      _animateCameraTo(diaryGeoCoordinatesList);
    } on InstantiationException catch (e) {
      print("Error: " + e.error.name);
    }

    setMessage(
        "Diary Entry from: " + gpxTrack.description + "\n" + "Hike Distance: " + distanceTravelled.toString() + " m");
  }

  void deleteDiaryEntry(int index) {
    Future<bool> isSuccess = gpxManager.deleteGPXTrack(index);
    isSuccess.then((value) => setMessage("Deleted entry: " + value.toString()));
  }

  void saveDiaryEntry() {
    // Permanently store the trip on the device.
    Future<bool> result = gpxManager.saveGPXTrack(gpxTrackWriter.track);

    result.then((value) => setMessage("Saved Hike: " + value.toString() + "."));
  }

  void onStartHikingButtonClicked() {
    clearMap();
    isHiking = true;
    moveMapToCurrentLocation = true;
    isGPXTrackLoaded = false;
    setMessage("Start Hike.");
    gpxTrackWriter = new GPXTrackWriter();
  }

  void onStopHikingButtonClicked() {
    clearMap();
    if (isHiking && !isGPXTrackLoaded) {
      saveDiaryEntry();
    } else {
      setMessage("Stopped.");
    }
    isHiking = false;
    moveMapToCurrentLocation = false;
    mapView.camera.cancelAnimations();
  }

  void enableOutdoorRasterLayer() {
    if (outdoorRasterLayer != null) {
      outdoorRasterLayer!.enable();
    }
  }

  void disableOutdoorRasterLayer() {
    if (outdoorRasterLayer != null) {
      outdoorRasterLayer!.disable();
    }
  }

  int getLengthOfGeoPolylineInMeters(GeoPolyline geoPolyline) {
    int length = 0;

    for (int i = 1; i < geoPolyline.vertices.length; i++) {
      length += geoPolyline.vertices.elementAt(i).distanceTo(geoPolyline.vertices.elementAt(i - 1)).toInt();
    }
    return length;
  }

  List<String> getMenuEntryKeys() {
    List<String> entryKeys = [];
    for (GPXTrack track in gpxManager.gpxDocument.tracks) {
      entryKeys.add(track.name);
    }
    return entryKeys;
  }

  List<String> getMenuEntryDescriptions() {
    List<String> entryDescriptions = [];

    for (GPXTrack track in gpxManager.gpxDocument.tracks) {
      entryDescriptions.add("Hike done on: " + track.description);
    }
    return entryDescriptions;
  }

  void setMessage(String message) {
    messageNotifier.updateMessage(message);
  }

  void _addMapPolyline(GeoPolyline geoPolyline) {
    clearMap();
    double widthInPixels = 20.0;
    Color polylineColor = const Color.fromARGB(0, 56, 54, 63);
    MapPolyline myPathMapPolyline;
    try {
      myPathMapPolyline = MapPolyline.withRepresentation(
          geoPolyline,
          MapPolylineSolidRepresentation(
              MapMeasureDependentRenderSize.withSingleSize(RenderSizeUnit.pixels, widthInPixels),
              polylineColor,
              LineCap.round));
      mapView.mapScene.addMapPolyline(myPathMapPolyline!);
    } on MapPolylineRepresentationInstantiationException catch (e) {
      print("MapPolylineRepresentation Exception:" + e.error.name);
      return;
    } on MapMeasureDependentRenderSizeInstantiationException catch (e) {
      print("MapMeasureDependentRenderSize Exception:" + e.error.name);
      return;
    }
  }

  MapPolyline _updateTravelledPath() {
    List<GeoCoordinates> geoCoordinatesList = gpxManager.getGeoCoordinatesList(gpxTrackWriter.track);
    if (geoCoordinatesList.length < 2) {
      return MapPolyline.withRepresentation(
          GeoPolyline([]),
          MapPolylineSolidRepresentation(
              MapMeasureDependentRenderSize.withSingleSize(RenderSizeUnit.pixels, 0),
              Colors.transparent,
              LineCap.round));
    }
    GeoPolyline geoPolyline;
    try {
      geoPolyline = new GeoPolyline(geoCoordinatesList);
      if (myPathMapPolyline == null) {
        _addMapPolyline(geoPolyline);
        return myPathMapPolyline!;
      }
      myPathMapPolyline!.geometry = geoPolyline;
    } on InstantiationErrorCode catch (e) {
      print("Error: " + e.name);
    }

    return myPathMapPolyline!;
  }

  void _animateCameraTo(List<GeoCoordinates> geoCoordinateList) {
    // We want to show the polyline fitting in the map view with an additional padding of 50 pixels.
    Point2D origin = new Point2D(50.0, 50.0);
    Size2D sizeInPixels = new Size2D(mapView.viewportSize.width - 100, mapView.viewportSize.height - 100);
    Rectangle2D mapViewport = new Rectangle2D(origin, sizeInPixels);

    // Untilt and unrotate the map.
    double bearing = 0;
    double tilt = 0;
    GeoOrientationUpdate geoOrientationUpdate = new GeoOrientationUpdate(bearing, tilt);

    // For very short polylines we want to have at least a distance of 100 meters.
    MapMeasure minDistanceInMeters = new MapMeasure(MapMeasureKind.distance, 100);

    MapCameraUpdate mapCameraUpdate =
        MapCameraUpdateFactory.lookAtPoints(geoCoordinateList, mapViewport, geoOrientationUpdate, minDistanceInMeters);

    // Create animation.
    Duration durationInSeconds = Duration(seconds: 3);
    MapCameraAnimation mapCameraAnimation =
        MapCameraAnimationFactory.createAnimationFromUpdateWithEasing(mapCameraUpdate, durationInSeconds, Easing(EasingFunction.inCubic));

    mapView.camera.startAnimation(mapCameraAnimation);
  }

  void _animateCameraToCurrentLocation(Location currentLocation) {
    GeoCoordinatesUpdate geoCoordinatesUpdate = GeoCoordinatesUpdate.fromGeoCoordinates(currentLocation.coordinates);
    Duration durationInSeconds = Duration(seconds: 3);
    MapMeasure distanceInMeters = new MapMeasure(MapMeasureKind.distance, 500);
    MapCameraAnimation animation =
        MapCameraAnimationFactory.flyToWithZoom(geoCoordinatesUpdate, distanceInMeters, 1, durationInSeconds);
    mapView.camera.startAnimation(animation);
  }

  void onDestroyOutdoorRasterLayer() {
    if (outdoorRasterLayer != null) {
      outdoorRasterLayer!.onDestroy();
    }
  }
}

File: examples/latest/navigate/ios/HikingDiary/HikingDiary/HikingApp.swift
/*
 * Copyright (C) 2022-2024 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import heresdk
import UIKit

// An app that tracks the user's location and allows to record the travelled path with a GPXTrackWriter.
class HikingApp: LocationDelegate {

    // A class to receive location events from the device.
    private let herePositioningProvider = HEREPositioningProvider()

    private let mapView: MapView
    private var myPathMapPolyline: MapPolyline?
    private var isHiking = false
    private var isGPXTrackLoaded = false
    private var gpxTrackWriter = GPXTrackWriter()
    private let gpxManager: GPXManager
    private let positioningVisualizer: HEREPositioningVisualizer
    private let outdoorRasterLayer: OutdoorRasterLayer
    private let messageTextView = UITextView()
    private var locationFilter: LocationFilterStrategy

    init(mapView: MapView) {
        self.mapView = mapView

        // A class to filter out undesired location signals.
        locationFilter = DistanceAccuracyLocationFilter()

        // A class to manage GPX operations.
        gpxManager = GPXManager(gpxDocumentFileName: "myGPXDocument.gpx")

        // A helper to show our current location and the raw location signals.
        positioningVisualizer = HEREPositioningVisualizer(mapView)

        // On top of the default map style, we allow to show a dedicated outdoor layer.
        outdoorRasterLayer = OutdoorRasterLayer(mapView: mapView)

        // Sets delegate to receive locations from HERE Positioning.
        herePositioningProvider.startLocating(locationDelegate: self, accuracy: .navigation)

        animateCameraToCurrentLocation()

        setupMessageView()
        setMessage("** Hiking Diary **")
    }

    func onStartHikingButtonClicked() {
        clearMap()
        isHiking = true
        isGPXTrackLoaded = false
        animateCameraToCurrentLocation()
        setMessage("Start Hike.")

        // Create a new GPXTrackWriter to record this trip.
        gpxTrackWriter = GPXTrackWriter()
    }

    func onStopHikingButtonClicked() {
        clearMap()

        if isHiking && isGPXTrackLoaded == false {
            saveDiaryEntry()
        } else {
            setMessage("Stopped.")
        }

        isHiking = false
    }

    func enableOutdoorRasterLayer() {
        outdoorRasterLayer.enable()
    }

    func disableOutdoorRasterLayer() {
        outdoorRasterLayer.disable()
    }

    // Conform to LocationDelegate protocol.
    func onLocationUpdated(_ location: heresdk.Location) {
        positioningVisualizer.updateLocationIndicator(location)

        if isHiking {
            // Here we visualize all incoming locations, unfiltered.
            positioningVisualizer.renderUnfilteredLocationSignals(location)
        }

        if isHiking && locationFilter.checkIfLocationCanBeUsed(location) {
            // Here we store only locations that are considered to be accurate enough.
            gpxTrackWriter.onLocationUpdated(location)

            // Visualize our hiking trip as an expanding polyline.
            let mapPolyline = updateTravelledPath()

            // Calculate the currently travelled distance.
            if let mapPolyline = mapPolyline {
                let distanceTravelled = getLengthOfGeoPolylineInMeters(mapPolyline.geometry)
                setMessage("Hike Distance: \(distanceTravelled) m")
            }
        }
    }
    
    private func updateTravelledPath() -> MapPolyline? {
        let geoCoordinatesList = gpxManager.getGeoCoordinatesList(track: gpxTrackWriter.track)

        if geoCoordinatesList.count < 2 {
            return nil
        }

        // We are sure that the number of vertices is greater than 1 (see above), so it will not crash.
        let geoPolyline = try! GeoPolyline(vertices: geoCoordinatesList)

        // Add polyline to the map, if instance is nil.
        guard let mapPolyline = myPathMapPolyline else {
            addMapPolyline(geoPolyline)
            return myPathMapPolyline
        }

        // Update the polyline shape that shows the travelled path of the user.
        mapPolyline.geometry = geoPolyline

        return mapPolyline
    }

    private func getLengthOfGeoPolylineInMeters(_ geoPolyLine: GeoPolyline) -> Int {
        var length = 0
        for i in 1...geoPolyLine.vertices.count - 1 {
            length =  length + Int(geoPolyLine.vertices[i].distance(to: geoPolyLine.vertices[i-1]))
        }
        return length
    }

    private func addMapPolyline(_ geoPolyline: GeoPolyline) {
        clearMap()
        let widthInPixels = 20.0
        let polylineColor = UIColor(red: 0, green: 0.56, blue: 0.54, alpha: 0.63)
        do {
            myPathMapPolyline =  try MapPolyline(geometry: geoPolyline,
                                                    representation: MapPolyline.SolidRepresentation(
                                                        lineWidth: MapMeasureDependentRenderSize(
                                                            sizeUnit: RenderSize.Unit.pixels,
                                                            size: widthInPixels),
                                                        color: polylineColor,
                                                        capShape: LineCap.round))
            
            mapView.mapScene.addMapPolyline(myPathMapPolyline!)
        } catch let error {
            fatalError("Failed to render MapPolyline. Cause: \(error)")
        }
    }

    private func clearMap() {
        if myPathMapPolyline != nil {
            mapView.mapScene.removeMapPolyline(myPathMapPolyline!)
            myPathMapPolyline = nil
        }
        positioningVisualizer.clearMap()
    }

    private func saveDiaryEntry() {
        // Permanently store the trip on the device.
        let result = gpxManager.saveGPXTrack(gpxTrackWriter.track)
        setMessage("Saved Hike: \(result).")
    }

    // Load the selected diary entry and show the polyline related to that hike.
    public func loadDiaryEntry(index: Int) {
        if isHiking {
            print("Stop hiking first.")
            return
        }

        isGPXTrackLoaded = true

        // Load the hiking trip.
        guard let gpxTrack = gpxManager.getGPXTrack(index: index) else {
            return
        }

        let diaryGeoCoordinatesList = gpxManager.getGeoCoordinatesList(track: gpxTrack)
        let diaryGeoPolyline = try! GeoPolyline(vertices: diaryGeoCoordinatesList)
        let distanceTravelled = getLengthOfGeoPolylineInMeters(diaryGeoPolyline)

        addMapPolyline(diaryGeoPolyline)
        animateCameraTo(diaryGeoCoordinatesList)

        setMessage("Diary Entry from: " + gpxTrack.description + "\n" +
                   "Hike Distance: \(distanceTravelled) m")
    }

    public func deleteDiaryEntry(index: Int) {
        let isSuccess = gpxManager.deleteGPXTrack(index: index)
        print("Deleted entry: \(isSuccess)")
    }

    public func getMenuEntryKeys() -> [String] {
        var entryKeys: [String] = []
        for track in gpxManager.gpxDocument.tracks {
            entryKeys.append(track.name)
        }
        return entryKeys
    }

    public func getMenuEntryDescriptions() -> [String] {
        var entryDescriptions: [String] = []
        for track in gpxManager.gpxDocument.tracks {
            entryDescriptions.append("Hike done on: " + track.description)
        }
        return entryDescriptions
    }

    private func animateCameraToCurrentLocation() {
        if let currentLocation = herePositioningProvider.getLastKnownLocation() {
            let geoCoordinatesUpdate = GeoCoordinatesUpdate(currentLocation.coordinates)
            let durationInSeconds = TimeInterval(3)
            let distanceInMeters = MapMeasure(kind: .distance, value: 500)
            let animation = MapCameraAnimationFactory.flyTo(target: geoCoordinatesUpdate,
                                                            zoom: distanceInMeters,
                                                            bowFactor: 1,
                                                            duration: durationInSeconds)
            mapView.camera.startAnimation(animation)
        }
    }

    private func animateCameraTo(_ geoCoordinateList: [GeoCoordinates]) {
        // We want to show the polyline fitting in the map view with an additional padding of 50 pixels.
        let origin:Point2D = Point2D(x: 50.0, y: 50.0)
        let sizeInPixels:Size2D = Size2D(width: mapView.viewportSize.width - 100, height: mapView.viewportSize.height - 100)
        let mapViewport = Rectangle2D(origin: origin, size: sizeInPixels)

        // Untilt and unrotate the map.
        let bearing: Double = 0
        let tilt: Double = 0
        let geoOrientationUpdate = GeoOrientationUpdate(bearing: bearing, tilt: tilt)

        // For very short polylines we want to have at least a distance of 100 meters.
        let minDistanceInMeters = MapMeasure(kind: .distance, value: 100)

        let mapCameraUpdate = MapCameraUpdateFactory.lookAt(geoCoordinateList,
                                                            viewRectangle: mapViewport,
                                                            orientation: geoOrientationUpdate,
                                                            measureLimit: minDistanceInMeters)

        // Create animation.
        let durationInSeconds = TimeInterval(3)
        let mapCameraAnimation = MapCameraAnimationFactory.createAnimation(from: mapCameraUpdate,
                                                                           duration: durationInSeconds,
                                                                           easing: Easing(EasingFunction.inCubic))
        mapView.camera.startAnimation(mapCameraAnimation)
    }

    // A permanent view to show scrollable text content.
    private func setupMessageView() {
        messageTextView.textColor = .white
        messageTextView.backgroundColor = UIColor(red: 0, green: 144 / 255, blue: 138 / 255, alpha: 1)
        messageTextView.layer.cornerRadius = 8
        messageTextView.isEditable = false
        messageTextView.textAlignment = NSTextAlignment.center
        messageTextView.font = .systemFont(ofSize: 14)
        messageTextView.frame = CGRect(x: 0, y: 0, width: mapView.frame.width * 0.6, height: mapView.frame.height * 0.085)
        messageTextView.center = CGPoint(x: mapView.frame.width * 0.5, y: mapView.frame.height * 0.9)

        UIView.transition(with: mapView, duration: 0.2, options: [.transitionCrossDissolve], animations: {
            self.mapView.addSubview(self.messageTextView)
        })
    }

    public func setMessage(_ message: String) {
        messageTextView.text = message
    }
}
